File: commands/buybitcoin.js
const Discord = require('discord.js');
const { Command } = require('./classes/command');
const { Bitcoin } = require('../classes/bitcoin');
const { format } = require('../utils/math');

class BuyBitcoin extends Command {
  constructor(client) {
    super(client, 'buybitcoin', 'buy bitcoin with mystic credits', [
      {
        name: 'amount',
        description: 'the amount of bitcoin to buy (negative to sell)',
        type: 10,
        required: true,
      },
    ]);
  }

  async run(interaction) {
    const bitcoin = new Bitcoin();
    const amount = interaction.options.getNumber('amount');
    const price = await bitcoin.getPrice();
    if (price === null) {
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#AA0000')
          .setTitle('Failed to get bitcoin price'),
        ],
      });
      return;
    }

    let credits = await this.client.db.user.getUserAttr(interaction.user.id, 'credits');
    let bitcoinAmount = await this.client.db.user.getUserAttr(interaction.user.id, 'bitcoin');

    if (price === null) {
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#AA0000')
          .setTitle('Failed to get bitcoin price'),
        ],
      });
      return;
    }

    if (amount === 0) {
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#AA0000')
          .setTitle('You bought gamebang\'s pp!'),
        ],
      });
    } else if (amount < 0) {
      if (bitcoinAmount < -amount) {
        await interaction.editReply({
          embeds: [new Discord.EmbedBuilder()
            .setColor('#AA0000')
            .setTitle('You don\'t have that much bitcoin to sell smh'),
          ],
        });
      } else {
        await this.client.db.user.addUserAttr(interaction.user.id, 'bitcoin', amount);
        await this.client.db.user.addUserAttr(interaction.user.id, 'credits', -amount * price);
        await this.client.db.user.addUserAttr(interaction.user.id, 'totalSoldAmount', -amount);
        await this.client.db.user.addUserAttr(interaction.user.id, 'totalSoldPrice', -amount * price);
        bitcoinAmount = await this.client.db.user.getUserAttr(interaction.user.id, 'bitcoin');
        credits = await this.client.db.user.getUserAttr(interaction.user.id, 'credits');
        const lastBoughtAmount = await this.client.db.user.getUserAttr(interaction.user.id, 'lastBoughtAmount');
        const lastBoughtPrice = await this.client.db.user.getUserAttr(interaction.user.id, 'lastBoughtPrice');
        await interaction.editReply({
          embeds: [new Discord.EmbedBuilder()
            .setColor('#00AA00')
            .setTitle(`Sold ${-amount} bitcoin for ${format(-amount * price)} mystic credits!`)
            .setDescription(`Current bitcoin price: ${format(price)}
Last bought price: ${format(lastBoughtPrice)} (${lastBoughtAmount} bitcoin)
Current bitcoin amount: ${bitcoinAmount}
Current mystic credits: ${format(credits)}`),
          ],
        });
      }
    } else if (credits < amount * price) {
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#AA0000')
          .setTitle('You\'re too poor to buy that much bitcoin smh'),
        ],
      });
    } else {
      await this.client.db.user.addUserAttr(interaction.user.id, 'bitcoin', amount);
      await this.client.db.user.addUserAttr(interaction.user.id, 'credits', -amount * price);
      await this.client.db.user.addUserAttr(interaction.user.id, 'totalBoughtAmount', amount);
      await this.client.db.user.addUserAttr(interaction.user.id, 'totalBoughtPrice', amount * price);
      await this.client.db.user.addUserAttr(interaction.user.id, 'lastBoughtAmount', amount);
      await this.client.db.user.addUserAttr(interaction.user.id, 'lastBoughtPrice', price);
      bitcoinAmount = await this.client.db.user.getUserAttr(interaction.user.id, 'bitcoin');
      credits = await this.client.db.user.getUserAttr(interaction.user.id, 'credits');
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#00AA00')
          .setTitle(`Bought ${amount} bitcoin for ${format(amount * price)} mystic credits!`)
          .setDescription(`Current bitcoin price: ${format(price)}
Current bitcoin amount: ${bitcoinAmount}
Current mystic credits: ${format(credits)}`),
        ],
      });
    }
  }
}

module.exports = BuyBitcoin;

File: src/commands/buybitcoin.rs
use crate::{Context, Error, utils, bitcoin};
use poise::serenity_prelude as serenity;

/// buy bitcoin with mystic credits
#[poise::command(slash_command)]
pub async fn buybitcoin(
    ctx: Context<'_>,
    #[description = "the amount of bitcoin to buy (negative to sell)"] amount: f64,
) -> Result<(), Error> {
    ctx.defer().await?;

    let price = match bitcoin::get_bitcoin_price().await {
        Ok(p) => p,
        Err(_) => {
            ctx.say("Failed to get bitcoin price").await?;
            return Ok(());
        }
    };

    let user_id = ctx.author().id.to_string();
    let mut db_user = ctx.data().db.get_user(&user_id).await?;

    if amount == 0.0 {
        ctx.say("You bought gamebang's pp!").await?;
    } else if amount < 0.0 {
        if db_user.bitcoin < -amount {
            ctx.say("You don't have that much bitcoin to sell smh").await?;
        } else {
            ctx.data().db.update_user_attr(&user_id, "bitcoin", amount, true).await?;
            ctx.data().db.update_user_attr(&user_id, "credits", -amount * price, true).await?;
            ctx.data().db.update_user_attr(&user_id, "total_sold_amount", -amount, true).await?;
            ctx.data().db.update_user_attr(&user_id, "total_sold_price", -amount * price, true).await?;
            
            db_user = ctx.data().db.get_user(&user_id).await?;
            
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .color(0x00AA00)
                    .title(format!("Sold {} bitcoin for {} mystic credits!", -amount, utils::format(-amount * price)))
                    .description(format!("Current bitcoin price: {}\nLast bought price: {} ({} bitcoin)\nCurrent bitcoin amount: {}\nCurrent mystic credits: {}", 
                        utils::format(price),
                        utils::format(db_user.last_bought_price),
                        db_user.last_bought_amount,
                        db_user.bitcoin,
                        utils::format(db_user.credits as f64)
                    )))
            ).await?;
        }
    } else if (db_user.credits as f64) < amount * price {
        ctx.say("You're too poor to buy that much bitcoin smh").await?;
    } else {
        ctx.data().db.update_user_attr(&user_id, "bitcoin", amount, true).await?;
        ctx.data().db.update_user_attr(&user_id, "credits", -amount * price, true).await?;
        ctx.data().db.update_user_attr(&user_id, "total_bought_amount", amount, true).await?;
        ctx.data().db.update_user_attr(&user_id, "total_bought_price", amount * price, true).await?;
        ctx.data().db.update_user_attr(&user_id, "last_bought_amount", amount, false).await?;
        ctx.data().db.update_user_attr(&user_id, "last_bought_price", price, false).await?;
        
        db_user = ctx.data().db.get_user(&user_id).await?;
        
        ctx.send(poise::CreateReply::default()
            .embed(serenity::CreateEmbed::new()
                .color(0x00AA00)
                .title(format!("Bought {} bitcoin for {} mystic credits!", amount, utils::format(amount * price)))
                .description(format!("Current bitcoin price: {}\nCurrent bitcoin amount: {}\nCurrent mystic credits: {}", 
                    utils::format(price),
                    db_user.bitcoin,
                    utils::format(db_user.credits as f64)
                )))
        ).await?;
    }

    Ok(())
}
